File: src/hard/MaxPathSum.java
package hard;


import common.TreeNode;

public class MaxPathSum {
    private int maxSum;
    public int maxPathSum(TreeNode root) {
        maxSum = Integer.MIN_VALUE;
        helper(root);
        return maxSum;
    }

    private int helper(TreeNode root) {
        if (root == null) return 0;
        int left = Math.max(helper(root.left), 0);
        int right = Math.max(helper(root.right), 0);
        // set max val is max left + max right + mid vs. last max
        maxSum = Math.max(maxSum, root.val + left + right);
        // return path to root
        return root.val + Math.max(left, right);
    }
}

File: csharp/LeetCode/Hard/MaxPathSum.cs
using System;
using LeetCode.DataStructures;

namespace LeetCode.Hard
{
    public partial class Solution
    {
        public int MaxPathSum(TreeNode root)
        {
            return this.MaxPathSumInternal1(root).Item2;
        }

        /// <summary>
        /// MaxPathSumInternal
        /// </summary>
        /// <param name="root"></param>
        /// <returns>(SumToRoot, SumFromSubTree)</returns>
        (int, int) MaxPathSumInternal1(TreeNode root)
        {
            if (root == null)
            {
                return (0, int.MinValue);
            }
            var (leftToRoot, maxSumFromLeft) = this.MaxPathSumInternal1(root.left);
            var (rightToRoot, maxSumFromRight) = this.MaxPathSumInternal1(root.right);

            var leftPath = Math.Max(leftToRoot, 0);
            var rightPath = Math.Max(rightToRoot, 0);
            var sumToRoot = root.val + Math.Max(leftPath, rightPath);
            var sumPassRoot = leftPath + rightPath + root.val;

            var maxSum = Math.Max(maxSumFromLeft, maxSumFromRight);
            maxSum = Math.Max(maxSum, sumPassRoot);
            return (sumToRoot, maxSum);
        }
    }
}
